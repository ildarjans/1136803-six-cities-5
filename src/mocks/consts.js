export const MOCK_OFFERS_COUNT = 100;
export const MOCK_REVIEWS_COUNT = 5;
export const AVATAR_URL = `https://avatars.dicebear.com/api`;
export const MAX_FEATURES = 3;
export const imageGallery = {
  MIN: 2,
  MAX: 8,
};

export const priceRange = {
  MIN: 19,
  MAX: 4999,
};
export const bedroomsRange = {
  MIN: 1,
  MAX: 6,
};
export const guestRange = {
  MIN: 1,
  MAX: 10,
};

export const avatarSprites = [
  `male`,
  `female`,
  `human`,
  `bottts`,
  `avataaars`,
  `gridy`,
];


export const names = [
  `James`,
  `John`,
  `Robert`,
  `William`,
  `Joseph`,
  `Charles`,
  `Thomas`,
  `Christopher`,
  `Anthony`,
  `Mark`,
  `Paul`,
  `Steven`,
  `Kenneth`,
  `Edward`,
  `Kevin`,
  `Ronald`,
  `Jason`,
  `Jeffrey`,
  `Gary`,
  `Nicholas`,
  `Stephen`,
  `Jonathan`,
  `Frank`,
  `Justin`,
  `Gregory`,
  `Samuel`,
  `Mary`,
  `Patricia`,
  `Elizabeth`,
  `Jessica`,
  `Sarah`,
  `Dorothy`,
  `Karen`,
  `Nancy`,
  `Sandra`,
  `Jennifer`,
  `Amanda`,
  `Rebecca`,
  `Sharon`,
  `Cynthia`,
  `Pamela`,
  `Katherine`,
  `Nicole`,
  `Samantha`,
  `Carolyn`,
];
export const lastNames = [
  `Smith`,
  `Brown`,
  `Johnson`,
  `Jones`,
  `Williams`,
  `Davis`,
  `Clark`,
  `White`,
  `Moore`,
  `Thompson`,
  `Adams`,
  `Thomas`,
  `Anderson`,
  `Lewis`,
  `Jackson`,
  `Green`,
  `Wood`,
  `Parker`,
  `Campbell`,
  `Robinson`,
  `Scott`,
  `Phillips`,
  `Turner`,
  `Carter`,
];

export const features = [
  `Wi-fi`,
  `Dishwasher`,
  `Parking`,
  `Washer`,
  `Elevator`,
  `Conditioner`,
  `Gym`,
  `Guide`,
  `Video games`,
  `Towels`,
  `Kitchen`,
  `Heating`,
];

export const descriptions = [
  `Lorem ipsum dolor sit amet, consectetur adipiscingâ€¦empor incididunt ut labore et dolore magna aliqua`,
  `Odio aenean sed adipiscing diam donec adipiscing tristique risus nec`,
  `Ut ornare lectus sit amet est placerat`,
  `Habitant morbi tristique senectus et`,
  `Ac tortor vitae purus faucibus ornare suspendisse sed nisi lacus`,
  `Tortor aliquam nulla facilisi cras fermentum odio eu feugiat pretium`,
  `Mattis enim ut tellus elementum`,
  `Justo donec enim diam vulputate ut pharetra sit amet aliquam`,
  `Elit ut aliquam purus sit amet luctus venenatis lectus`,
  `Eget velit aliquet sagittis id`,
  `Mi sit amet mauris commodo quis imperdiet massa`,
  `Ullamcorper a lacus vestibulum sed arcu non odio`,
  `Elementum eu facilisis sed odio morbi quis commodo odio`,
  `Et malesuada fames ac turpis egestas integer eget aliquet nibh`,
  `Pharetra massa massa ultricies mi quis hendrerit dolor`,
  `Eu turpis egestas pretium aenean pharetra magna ac`,
  `Dictum fusce ut placerat orci`,
  `Nisi vitae suscipit tellus mauris a diam maecenas sed`,
  `Mauris sit amet massa vitae tortor condimentum lacinia quis vel`,
  `Id faucibus nisl tincidunt eget nullam non nisi est sit.`,
];

export const titles = [
  `Beautiful & luxurious apartment at great location`,
  `Couples particularly like this location `,
  `Big company apartment`,
  `Nearest airport & transport station`,
  `Quite place for zen vacations`,
  `Best lobby bar & Michelin restaurant`,
  `Bed and breakfast. Best offer!`,
];
