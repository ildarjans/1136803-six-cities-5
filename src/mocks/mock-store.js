import configureStore from "redux-mock-store";
import {
  mockCityOffers,
  mockFavoritesOffers,
  mockOffers,
  mockReviews,
  mockUser
} from "./mock-data";

const createStore = configureStore();

const mockStoreTemplate = {
  OFFERS: {
    offers: mockOffers,
    cityOffers: mockCityOffers,
    loading: true,
    error: null
  },
  REVIEWS: {
    reviews: mockReviews,
    loading: false,
    error: null,
    posting: false
  },
  NEAR_OFFERS: {
    nearOffers: mockCityOffers.AMSTERDAM.slice(0, 3),
    loading: false,
    error: null
  },
  PROCESS: {
    activeCity: `AMSTERDAM`,
    hoveredOfferId: ``,
    sortType: ``
  },
  USER: {
    authorizationStatus: `NO_AUTHORIZED`,
    waitingAuthorizationResponse: false,
    error: null,
    user: mockUser
  },
  FAVORITES: {
    favoriteCityOffers: mockFavoritesOffers,
    updating: false,
    loading: false,
    error: null,
  }
};

export const mockStore = createStore(mockStoreTemplate);
