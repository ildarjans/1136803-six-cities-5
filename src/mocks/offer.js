import {
  generateId,
  getLocationCoords,
  getRandomArrayElement,
  getRandomArrayElements,
  getRandomAvatar,
  getRandomFullName,
  getRandomImageGallery,
  getRandomInteger,
} from "./utils";
import {offerTypes, cities} from "../const";
import {
  descriptions,
  features,
  BEDROOMS_RANGE,
  GUEST_RANGE,
  MAX_FEATURES,
  PRICE_RANGE,
  titles,
} from "./consts";

export function getOffer() {
  const city = getRandomArrayElement(cities);
  return {
    city,
    id: generateId(),
    images: getRandomImageGallery(),
    premium: Boolean(getRandomInteger(1)),
    price: getRandomInteger(PRICE_RANGE.MAX, PRICE_RANGE.MIN),
    title: getRandomArrayElement(titles),
    type: getRandomArrayElement(offerTypes),
    rating: getRandomInteger(5),
    description: getRandomArrayElement(descriptions),
    bedrooms: getRandomInteger(BEDROOMS_RANGE.MAX, BEDROOMS_RANGE.MIN),
    guests: getRandomInteger(GUEST_RANGE.MAX, GUEST_RANGE.MIN),
    features: getRandomArrayElements(features, MAX_FEATURES),
    owner: {
      avatar: getRandomAvatar(),
      name: getRandomFullName(),
      super: Boolean(getRandomInteger(1)),
    },
    coords: getLocationCoords(city)
  };
}

