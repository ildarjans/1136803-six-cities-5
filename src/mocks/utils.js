import {
  AVATAR_URL,
  AVATAR_SPRITES,
  LAST_NAMES,
  NAMES,
  IMAGE_GALLERY,
} from "./consts";

export function getRandomInteger(max, min = 0) {
  return min + Math.floor(Math.random() * (max + 1 - min));
}

export function getRandomArrayElements(arr, count) {
  if (arr.length < count) {
    return undefined;
  }
  const resultArray = [];
  while (resultArray.length < count) {
    const randElement = arr[getRandomInteger(arr.length - 1)];
    if (!resultArray.includes(randElement)) {
      resultArray.push(randElement);
    }
  }
  return resultArray;
}

export function getRandomArrayElement(arr) {
  if (!arr && arr.length === 0) {
    return undefined;
  }
  return arr[getRandomInteger(arr.length - 1)];
}

export function getRandomFullName() {
  return (
    `${getRandomArrayElement(NAMES)} ${getRandomArrayElement(LAST_NAMES)}`
  );
}

export function getRandomAvatar() {
  return `${AVATAR_URL}/${getRandomArrayElement(AVATAR_SPRITES)}/${Math.random()}.svg`;
}

export function getRandomImageGallery() {
  const gallerySize = getRandomInteger(IMAGE_GALLERY.MAX, IMAGE_GALLERY.MIN);
  return Array(gallerySize).fill().map(getRandomPhoto);
}

function getRandomPhoto() {
  return `http://picsum.photos/248/152?r=${Math.random()}`;
}

export function generateId() {
  return getRandomInteger(Date.now()).toString(16);
}
