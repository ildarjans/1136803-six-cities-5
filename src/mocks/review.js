import {
  getRandomArrayElement,
  getRandomAvatar,
  getRandomFullName,
  getRandomInteger,
} from "./utils";
import {descriptions} from "./consts";

export function getReview() {
  return {
    avatar: getRandomAvatar(),
    name: getRandomFullName(),
    date: new Date(),
    rating: getRandomInteger(5),
    review: getRandomArrayElement(descriptions),
  };
}

