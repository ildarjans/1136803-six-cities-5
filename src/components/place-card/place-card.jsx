import React from "react";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {Settings} from "../../const";
import {debounce, getRatingWidth} from "../../utils";
import {offerPropTypes} from "../../prop-types/offer";
import {actionCreator} from "../../store/actions";

export const PlaceCardComponent = (props) => {
  const {offer, classNameArticle, classNameWrapper, onActiveOfferChange} = props;
  const handleDebouncedActiveOfferChange = debounce(
      onActiveOfferChange,
      Settings.MAP_PIN_DEBOUNCE_DELAY
  );
  return (
    <article
      className={`${classNameArticle} place-card`}
      onMouseOver={() => {
        handleDebouncedActiveOfferChange(offer.id);
      }}
      onMouseOut={() => {
        handleDebouncedActiveOfferChange(``);
      }}
    >
      <div className={`${classNameWrapper} place-card__image-wrapper`}>
        <Link to={`/offer/${offer.id}`}>
          <img
            className="place-card__image"
            src={offer.previewImage}
            width="260"
            height="200"
            alt={offer.title}
          />
        </Link>

      </div>
      <div className="place-card__info">
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{offer.price}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>
          <button className="place-card__bookmark-button button" type="button">
            <svg className="place-card__bookmark-icon" width="18" height="19">
              <use xlinkHref="#icon-bookmark"/>
            </svg>
            <span className="visually-hidden">To bookmarks</span>
          </button>
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: `${getRatingWidth(offer.rating)}%`}}/>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 className="place-card__name">
          <Link to={`/offer/${offer.id}`}>{offer.title}</Link>
        </h2>
        <p className="place-card__type">{offer.type}</p>
      </div>
    </article>
  );

};

PlaceCardComponent.propTypes = {
  offer: offerPropTypes.isRequired,
  classNameArticle: PropTypes.string.isRequired,
  classNameWrapper: PropTypes.string.isRequired,
  onActiveOfferChange: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  onActiveOfferChange: (id) => {
    dispatch(actionCreator.changeActiveOfferId(id));
  }
});

export const PlaceCard = connect(null, mapDispatchToProps)(PlaceCardComponent);
