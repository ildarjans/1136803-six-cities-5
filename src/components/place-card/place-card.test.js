import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";
import {mockCityOffers} from "../../mocks/mock-data";
import {PlaceCard} from "./place-card";

const PLACE_CARD_OPTION = {
  PLACE_LIST: {
    CLASSNAME_ARTICLE: `cities__place-card`,
    CLASSNAME_WRAPPER: `cities__place-card`,
    IMAGE_WIDTH: 260,
    IMAGE_HEIGHT: 200,
  },
  NEAR_LIST: {
    CLASSNAME_ARTICLE: `near-places__card`,
    CLASSNAME_WRAPPER: `near-places__image-wrapper`,
    IMAGE_WIDTH: 260,
    IMAGE_HEIGHT: 200,
  },
  FAVORITE_LIST: {
    CLASSNAME_ARTICLE: `favorites__card`,
    CLASSNAME_WRAPPER: `favorites__image-wrapper`,
    IMAGE_WIDTH: 150,
    IMAGE_HEIGHT: 110,
  },
};

const noop = () => {};

jest.mock(`../favorite-button/favorite-button`);

describe(`Should PlaceCard rendered correctly`, () => {
  it(`(1) render Favorites list implementation`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <PlaceCard
              offer={mockCityOffers.AMSTERDAM[0]}
              options={PLACE_CARD_OPTION.FAVORITE_LIST}
            />
          </BrowserRouter>
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it(`(2) render Near list implementation`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <PlaceCard
              offer={mockCityOffers.AMSTERDAM[0]}
              options={PLACE_CARD_OPTION.NEAR_LIST}
              onActiveOfferChange={noop}
            />
          </BrowserRouter>
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it(`(3) render Places list implementation`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <PlaceCard
              offer={mockCityOffers.AMSTERDAM[0]}
              options={PLACE_CARD_OPTION.PLACE_LIST}
              renderChild={noop}
              onActiveOfferChange={noop}
            />
          </BrowserRouter>
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
