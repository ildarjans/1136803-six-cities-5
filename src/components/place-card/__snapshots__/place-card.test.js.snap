// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should PlaceCard rendered correctly (1) render Favorites list implementation 1`] = `
<article
  className="favorites__card place-card"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="favorites__image-wrapper place-card__image-wrapper"
  >
    <a
      href="/offer/73"
      onClick={[Function]}
    >
      <img
        alt="Big company apartment"
        className="place-card__image"
        height={110}
        src="http://picsum.photos/248/152?r=0.16844568825898554"
        width={150}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          2523
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "20%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="/offer/73"
        onClick={[Function]}
      >
        Big company apartment
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      apartment
    </p>
  </div>
</article>
`;

exports[`Should PlaceCard rendered correctly (2) render Near list implementation 1`] = `
<article
  className="near-places__card place-card"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="near-places__image-wrapper place-card__image-wrapper"
  >
    <a
      href="/offer/73"
      onClick={[Function]}
    >
      <img
        alt="Big company apartment"
        className="place-card__image"
        height={200}
        src="http://picsum.photos/248/152?r=0.16844568825898554"
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          2523
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "20%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="/offer/73"
        onClick={[Function]}
      >
        Big company apartment
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      apartment
    </p>
  </div>
</article>
`;

exports[`Should PlaceCard rendered correctly (3) render Places list implementation 1`] = `
<article
  className="cities__place-card place-card"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="cities__place-card place-card__image-wrapper"
  >
    <a
      href="/offer/73"
      onClick={[Function]}
    >
      <img
        alt="Big company apartment"
        className="place-card__image"
        height={200}
        src="http://picsum.photos/248/152?r=0.16844568825898554"
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          2523
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "20%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="/offer/73"
        onClick={[Function]}
      >
        Big company apartment
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      apartment
    </p>
  </div>
</article>
`;
